Framework/Algorithms/src/CalculatePlaczekSelfScattering.cpp

```diff
@@ -75,35 +74,40 @@ std::map<std::string, std::string> CalculatePlaczekSelfScattering::validateInput
   return issues;
 }
 
+//----------------------------------------------------------------------------------------------
+double CalculatePlaczekSelfScattering::getPackingFraction(const API::MatrixWorkspace_const_sptr &ws) {
+  // get a handle to the material
+  const auto &material = ws->sample().getMaterial();
+
+  // default value is packing fraction
+  double packingFraction = material.packingFraction();
+
+  // see if the user thinks the material wasn't setup right
+  const double crystalDensity = getProperty("CrystalDensity");
+  if (crystalDensity > 0.) {
+    // assume that the number density set in the Material is the effective number density
+    packingFraction = material.numberDensity() / crystalDensity;
+  }
+
+  return packingFraction;
+}
+
 //----------------------------------------------------------------------------------------------
 /** Execute the algorithm.
  */
 void CalculatePlaczekSelfScattering::exec() {
   const API::MatrixWorkspace_sptr inWS = getProperty("InputWorkspace");
   const API::MatrixWorkspace_sptr incidentWS = getProperty("IncidentSpecta");
-  constexpr double factor = 1.0 / 1.66053906660e-27;       // atomic mass unit-kilogram relationship
-  constexpr double neutronMass = factor * 1.674927471e-27; // neutron mass
-  // get sample information : mass, total scattering length, and concentration
-  // of each species
-  auto atomSpecies = getSampleSpeciesInfo(inWS);
-  // calculate summation term w/ neutron mass over molecular mass ratio
 
-  auto sumLambda = [&neutronMass](double sum, auto &atom) {
-    return sum + atom.second["concentration"] * atom.second["bSqrdBar"] * neutronMass / atom.second["mass"];
-  };
-  double summationTerm = std::accumulate(atomSpecies.begin(), atomSpecies.end(), 0.0, sumLambda);
+  // calculate summation term w/ neutron mass over molecular mass ratio
+  const double summationTerm = calculateSummationTerm(inWS->sample().getMaterial());
+  const double packingFraction = getPackingFraction(inWS);
 
-  double numberDensity = inWS->sample().getMaterial().numberDensity();
-  double crystalDensity = getProperty("CrystalDensity");
-  double densityRatio = 1.0;
-  if (crystalDensity > 0) {
-    densityRatio = numberDensity / crystalDensity;
-  }
   // get incident spectrum and 1st derivative
   const MantidVec xLambda = incidentWS->readX(0);
   const MantidVec incident = incidentWS->readY(0);
   const MantidVec incidentPrime = incidentWS->readY(1);
-  double dx = (xLambda[1] - xLambda[0]) / 2.0;
+  const double dx = (xLambda[1] - xLambda[0]) / 2.0; // assume constant bin width
```

Comment:
Just a comment here, if someone pass in a workspace with a spectrum of 1 bin, we will have segmentation fault here.

Reply:
Ouch!